src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp

```diff
@@ -0,0 +1,227 @@
+    ControllerScriptEngineBase::shutdown();
+}
+
+bool ControllerScriptEngineLegacy::handleIncomingData(const QByteArray& data) {
+    // This function is called from outside the controller engine, so we can't
+    // use VERIFY_OR_DEBUG_ASSERT here
+    if (!m_pJSEngine) {
+        return false;
+    }
+
+    QJSValueList args;
+    args << m_pJSEngine->toScriptValue(data);
+    args << QJSValue(data.size());
+
+    for (const QJSValue& function : m_incomingDataFunctions) {
+        ControllerScriptEngineBase::executeFunction(function, args);
+    }
+
+    return true;
+}
+
+bool ControllerScriptEngineLegacy::evaluateScriptFile(const QFileInfo& scriptFile) {
+    VERIFY_OR_DEBUG_ASSERT(m_pJSEngine) {
+        return false;
+    }
+
+    if (!scriptFile.exists()) {
+        qWarning() << "File does not exist:"
+                   << scriptFile.absoluteFilePath();
+        return false;
+    }
+
+    // If the script is invalid, it should be watched so the user can fix it
+    // without having to restart Mixxx. So, add it to the watcher before
+    // evaluating it.
+    if (!m_fileWatcher.addPath(scriptFile.absoluteFilePath())) {
+        qWarning() << "Failed to watch script file" << scriptFile.absoluteFilePath();
+    };
+
+    qDebug() << "Loading"
+             << scriptFile.absoluteFilePath();
+
+    // Read in the script file
+    QString filename = scriptFile.absoluteFilePath();
+    QFile input(filename);
+    if (!input.open(QIODevice::ReadOnly)) {
+        qWarning() << QString(
+                "Problem opening the script file: %1, "
+                "error # %2, %3")
+                              .arg(filename,
+                                      QString::number(input.error()),
+                                      input.errorString());
+        // Set up error dialog
+        ErrorDialogProperties* props = ErrorDialogHandler::instance()->newDialogProperties();
+        props->setType(DLG_WARNING);
+        props->setTitle(tr("Controller Mapping File Problem"));
+        props->setText(tr("The mapping for controller \"%1\" cannot be opened.")
+                               .arg(m_pController->getName()));
+        props->setInfoText(
+                tr("The functionality provided by this controller mapping will "
+                   "be disabled until the issue has been resolved."));
+
+        // We usually don't translate the details field, but the cause of
+        // this problem lies in the user's system (e.g. a permission
+        // issue). Translating this will help users to fix the issue even
+        // when they don't speak english.
+        props->setDetails(tr("File:") + QStringLiteral(" ") + filename +
+                QStringLiteral("\n") + tr("Error:") + QStringLiteral(" ") +
+                input.errorString());
+
+        // Ask above layer to display the dialog & handle user response
+        ErrorDialogHandler::instance()->requestErrorDialog(props);
+        return false;
+    }
+
+    QString scriptCode;
```

Comment:
Use operator + (QStringBuilder) for all string concatenations

Reply:
Why? I find this easier to read and it works fine.

Reply:
QStringBuilder avoids intermediate allocations by resolving the concatenations at compile time. Template meta-programming magic.

Reply:
Okay, I'll change it. But squeezing a few nanoseconds from legacy code isn't particularly fruitful.